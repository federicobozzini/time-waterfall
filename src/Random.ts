import seedrandom from 'seedrandom';

export class Random {
  private prng: seedrandom.PRNG;

  constructor(seed?: string) {
    this.prng = seedrandom(seed);
  }

  public random(min: number, max: number): number {
    return this.prng.quick() * (max - min) + min;
  }
}
