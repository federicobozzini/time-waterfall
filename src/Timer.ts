export type OnTickCallback = (interval: number) => void;

export const MS_INTERVAL = 15;

export class Timer {

    private readonly cbs: OnTickCallback[] = [];

    public start() {
        this.refresh();
    }

    public onTick(cb: OnTickCallback): void {
        this.cbs.push(cb);
    }

    public tick(interval: number = MS_INTERVAL): void {
        this.cbs.forEach(cb => cb(interval));
    } 

    private refresh(): void {
        setTimeout(() => {
            this.cbs.forEach(cb => cb(MS_INTERVAL));
            this.refresh();
        }, MS_INTERVAL);
    }

}