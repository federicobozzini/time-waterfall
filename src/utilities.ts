export const times = (n: number, f: () => void) => {
    while(n-->0) {
        f();
    }
};