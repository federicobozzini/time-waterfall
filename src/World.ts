import { Timer } from "./Timer";

type OnElementsUpdate = (elements: Element[]) => void;

export interface Element {
    x: number;
    y: number;
    angle: number;
    value: number; /* 1 to 9 */
    size: number;
}


const l0 = {
    x: 95, 
    y: 40,
    angle: -30,
    value: 8,
    size: 12,
};
const l1 = {
    x: 40, 
    y: 90,
    angle: 5,
    value: 4,
    size: 8,
};
const l2 = {
    x: 20, 
    y: 15,
    angle: -70,
    value: 3,
    size: 22,
};
const l3 = {
    x: 85, 
    y: 75,
    angle: 10,
    value: 1,
    size: 8,
};

const yVel = 10;
const yMax = 130;

export class World {

    private elements = [l0, l1, l2, l3];
    private updateCbs: OnElementsUpdate[] = [];

    constructor(timer: Timer) {
        timer.onTick((interval: number) => {
            this.update(interval);
        });
    }

    public onUpdate(cb: OnElementsUpdate) {
        this.updateCbs.push(cb);
    }
    
    public getElements(): Element[] {
        return this.elements;
    }

    public removeCb(cb: OnElementsUpdate) {
        this.updateCbs = this.updateCbs.filter(cb1 => cb1 !== cb);
    }

    private update(interval:number) {
        this.elements = this.elements.map(e => 
            ({
                ...e, 
                y: e.y + yVel * (interval / 1000)
            })).filter(e => e.y < yMax);
        this.updateCbs.forEach(cb => cb(this.elements));
    }

}