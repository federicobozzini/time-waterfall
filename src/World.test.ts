import { World } from './World';
import { Timer } from './Timer';
import { times } from './utilities';
import { Random } from './Random';

let world: World;
let random: Random;
let timer: Timer;

describe('World', () => {
  beforeEach(() => {
    timer = new Timer(0);
    random = new Random('test');
    world = new World(timer, random);
  });

  test('fires onUpdate event on tick', () => {
    const n = 10;

    const cb = jest.fn();
    world.onUpdate(cb);

    times(n, () => timer.tick());

    expect(cb).toHaveBeenCalledTimes(n);
  });

  test('elements should change', () => {
    const ns = [0, 10, 100, 500, 2000, 5000, 10000];

    ns.forEach((n) => {
      times(n, () => timer.tick());

      const finalElements = world.getElements();

      expect(finalElements).toMatchSnapshot();
    });
  });
});
