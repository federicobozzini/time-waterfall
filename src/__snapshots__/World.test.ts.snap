// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`World elements should change 1`] = `Array []`;

exports[`World elements should change 2`] = `
Array [
  Object {
    "angle": -165.21320178285245,
    "angleVel": -35.49434378743172,
    "size": 15.289891721215099,
    "value": 2.0092316555820884,
    "valueVel": 0.21545813637785616,
    "x": 53.77714571157936,
    "xVel": -0.269928096793592,
    "y": 1.362047513346188,
    "yVel": 12.971881079487503,
  },
]
`;

exports[`World elements should change 3`] = `
Array [
  Object {
    "angle": -218.4547174640006,
    "angleVel": -35.49434378743172,
    "size": 15.289891721215099,
    "value": 1.6860444510153005,
    "valueVel": 0.21545813637785616,
    "x": 53.37225356638894,
    "xVel": -0.269928096793592,
    "y": 20.819869132577413,
    "yVel": 12.971881079487503,
  },
  Object {
    "angle": -40.0234843287617,
    "angleVel": -109.35166077688336,
    "size": 15.149289197288454,
    "value": 1.818579313322897,
    "valueVel": 2.8275511085987093,
    "x": 34.66990695750335,
    "xVel": 4.157810208387673,
    "y": 42.14579240852033,
    "yVel": 30.54042928153649,
  },
  Object {
    "angle": -156.6810995324517,
    "angleVel": -97.62650018557906,
    "size": 6.892302706371993,
    "value": 0.6029951854241073,
    "valueVel": 3.245461960043758,
    "x": 64.99528671470003,
    "xVel": -0.9512910833582282,
    "y": 8.247276416961103,
    "yVel": 28.937811989337206,
  },
]
`;

exports[`World elements should change 4`] = `
Array [
  Object {
    "angle": -266.6746498388238,
    "angleVel": -36.73896146938205,
    "size": 6.282950809691101,
    "value": 1.1595370256066782,
    "valueVel": 1.5156523545272647,
    "x": 79.62382787028824,
    "xVel": 0.47711201291531324,
    "y": 79.85008671564533,
    "yVel": 19.789364737458527,
  },
  Object {
    "angle": -26.967077577486634,
    "angleVel": 50.58165863156319,
    "size": 16.044193344656378,
    "value": 0.6367253861979095,
    "valueVel": 2.9961409229319544,
    "x": 56.27094857295972,
    "xVel": 2.4826386608183384,
    "y": 13.524224523184815,
    "yVel": 23.11833251826465,
  },
]
`;

exports[`World elements should change 5`] = `
Array [
  Object {
    "angle": 100.571814166009,
    "angleVel": 52.96269737184048,
    "size": 15.918192766606808,
    "value": 1.2469896338302742,
    "valueVel": 0.3757418837398291,
    "x": 66.36058797982028,
    "xVel": -2.9544910630211234,
    "y": 81.11453728459824,
    "yVel": 22.253645345568657,
  },
  Object {
    "angle": -86.57665416151308,
    "angleVel": -55.11049501597881,
    "size": 21.78380083758384,
    "value": 2.4686982887447417,
    "valueVel": 0.8553942560683936,
    "x": 77.82247395560259,
    "xVel": -1.2729493752121925,
    "y": 62.707205129884755,
    "yVel": 18.83099253149703,
  },
  Object {
    "angle": -95.25253353640417,
    "angleVel": -38.07477258145809,
    "size": 6.87708405777812,
    "value": 0.9349755195435304,
    "valueVel": 0.5170196841936558,
    "x": 35.46382457204176,
    "xVel": 5.195489217527211,
    "y": 40.98408845392979,
    "yVel": 13.661362817976624,
  },
  Object {
    "angle": -197.16910644136314,
    "angleVel": -81.93586707115173,
    "size": 12.848742100875825,
    "value": 3.773787420713776,
    "valueVel": 1.1923307996708898,
    "x": 54.45369487096592,
    "xVel": 5.528774888254702,
    "y": 99.45110984149848,
    "yVel": 46.69066189741716,
  },
  Object {
    "angle": -67.77316603325303,
    "angleVel": -66.20142146945,
    "size": 15.791896470822394,
    "value": 4.09667512946808,
    "valueVel": 2.286059636482969,
    "x": 82.38758779662635,
    "xVel": -5.161946981213987,
    "y": 36.28338720353324,
    "yVel": 41.70504276268184,
  },
  Object {
    "angle": 138.3397817540915,
    "angleVel": -31.66628010571003,
    "size": 7.657657964620739,
    "value": 6.0328196217357135,
    "valueVel": 2.3447727669961753,
    "x": 49.20257927967697,
    "xVel": 1.324142170138657,
    "y": 13.475160997824275,
    "yVel": 47.28126665903255,
  },
  Object {
    "angle": 131.42308366494245,
    "angleVel": -93.06689573451877,
    "size": 22.53528465796262,
    "value": 7.277810580169784,
    "valueVel": 3.541555866971612,
    "x": 15.561488375919867,
    "xVel": 2.733618135564029,
    "y": 1.0834121309872715,
    "yVel": 10.31821077130735,
  },
  Object {
    "angle": 72.47431498160586,
    "angleVel": -96.41944421455264,
    "size": 13.359900433570147,
    "value": 4.197495382126188,
    "valueVel": 2.9224481535609814,
    "x": 99.70979543026537,
    "xVel": 5.3274082615971565,
    "y": 0.3953779986617155,
    "yVel": 26.35853324411437,
  },
]
`;

exports[`World elements should change 6`] = `
Array [
  Object {
    "angle": 262.47897048164026,
    "angleVel": 27.900628242641687,
    "size": 15.429372511804104,
    "value": 1.501614136375997,
    "valueVel": 1.082333928765729,
    "x": 33.935035647618534,
    "xVel": -1.5021549370139837,
    "y": 95.12327731770891,
    "yVel": 20.196024908218533,
  },
  Object {
    "angle": 363.1510221194476,
    "angleVel": 117.37743727862835,
    "size": 22.513529579155147,
    "value": 2.27635413161921,
    "valueVel": 1.4902154235634952,
    "x": 103.84040494558784,
    "xVel": 1.850053514353931,
    "y": 62.999912713018816,
    "yVel": 28.571388985496014,
  },
  Object {
    "angle": 28.367694643512653,
    "angleVel": -23.19046899676323,
    "size": 11.20789027120918,
    "value": 3.196742716929851,
    "valueVel": 1.226729407440871,
    "x": 52.371188804656526,
    "xVel": -4.869590610265732,
    "y": 21.2010964295827,
    "yVel": 12.39830200560391,
  },
  Object {
    "angle": 25.6540492060596,
    "angleVel": -39.95118381455541,
    "size": 11.338602618779987,
    "value": 7.045660448685554,
    "valueVel": 1.623506988538429,
    "x": 84.86309334356363,
    "xVel": 0.05775144696235657,
    "y": 50.66551247471938,
    "yVel": 45.0360110886395,
  },
  Object {
    "angle": -178.6786413654686,
    "angleVel": -1.699121743440628,
    "size": 15.087324963882565,
    "value": 6.030010543771091,
    "valueVel": 2.143375358171761,
    "x": 30.515126408934606,
    "xVel": -5.29995778016746,
    "y": 5.962578777037562,
    "yVel": 49.68815647531301,
  },
  Object {
    "angle": -156.01511211600155,
    "angleVel": 53.746806643903255,
    "size": 19.10554161434993,
    "value": 6.551468036628796,
    "valueVel": 0.6957631616853177,
    "x": 37.84743474163115,
    "xVel": -3.9641669243574142,
    "y": 0.794018121866975,
    "yVel": 10.586908291559666,
  },
]
`;

exports[`World elements should change 7`] = `
Array [
  Object {
    "angle": 74.04798576179891,
    "angleVel": 98.38972393423319,
    "size": 19.05685656191781,
    "value": 6.39619534834849,
    "valueVel": 3.2465580724179746,
    "x": 6.821552326441768,
    "xVel": -5.835196950472891,
    "y": 20.203549530354785,
    "yVel": 27.48782249027863,
  },
  Object {
    "angle": 180.5991956569249,
    "angleVel": 10.428390204906464,
    "size": 22.669884414877743,
    "value": 6.764507469766782,
    "valueVel": 1.51498120483011,
    "x": 17.708461387693024,
    "xVel": -1.7012083986774087,
    "y": 10.977077074642304,
    "yVel": 15.570322091691196,
  },
  Object {
    "angle": 58.113469601422594,
    "angleVel": 90.09509917348623,
    "size": 7.158290677238256,
    "value": 7.861664724279194,
    "valueVel": 2.5632231157738716,
    "x": 91.39430201895539,
    "xVel": 1.158676665276289,
    "y": 10.04816882561892,
    "yVel": 41.86737010674551,
  },
  Object {
    "angle": 113.22235981654373,
    "angleVel": -95.76212709769607,
    "size": 7.800996467471123,
    "value": 3.5010153127810932,
    "valueVel": 1.2661140424665063,
    "x": 32.731386246765055,
    "xVel": -4.294558455236256,
    "y": 3.761392316431738,
    "yVel": 25.07594877621159,
  },
]
`;
